numpy/lib/function_base.py

```diff
@@ -1752,10 +1762,73 @@ def diff(a, n=1, axis=-1):
     slice1 = tuple(slice1)
     slice2 = tuple(slice2)
     if n > 1:
-        return diff(a[slice1]-a[slice2], n-1, axis=axis)
-    else:
+        return diff(a[slice1]-a[slice2], n-1, axis=axis, 
+            to_begin=to_begin, to_end=to_end)
+
+    # if nothing to add to either side, fast track the result
+    elif to_begin is None and to_end is None:
         return a[slice1]-a[slice2]
 
+    else:
+        # compute the length of the diff'd portion
+        # force length to be non negative
+        l_diff = a.shape[axis] - 1
+        if l_diff < 0:
+            l_diff = 0
+
+        # make to_end a 1D array
+        if to_end is None:
+            l_end = 0
+        else:
+            to_end = np.atleast_1d(to_end)
+            if to_end.ndim == 1:
+                l_end = len(to_end)
```

Comment:
Are you sure that is true for a 3-d array where you take the diff on axis=1 and have a 1-dimensional `to_end` with more than 1 element?
```
a = np.zeros((5,4,3))
to_end = np.array([1,2])
a[:, 2:, :] = to_end
# ValueError: could not broadcast input array from shape (2) into shape (5,2,3)
to_end.shape = (2, 1)
a[:, 2:, :] = to_end
# works
```